### PR TITLE
DIRECTOR: Debugger + Chop Suey fixes

### PR DESCRIPTION
**Pull request by moralrecordings**

### CODE REVIEW

File: engines/director/debugger.cpp [hunk 1]
```diff
@@ -329,6 +333,60 @@ bool Debugger::cmdBacktrace(int argc, const char **argv) {
 bool Debugger::cmdDisasm(int argc, const char **argv) {
 	Lingo *lingo = g_director->getLingo();
 	if (argc == 2) {
+		if (Common::String(argv[1]).equals("all")) {
```
Comment (sev-) [inline engines/director/debugger.cpp:336]: Python detected :) It is more readable, but I'd write `if (!strcmp(argv[1], "all"))`

File: engines/director/score.cpp [hunk 2]
```diff
@@ -488,6 +498,10 @@ void Score::update() {
 			} else if (tempo == 134) {
 				// Wait for sound channel 2
 				_waitForChannel = 2;
+
+			} else if (tempo >= 136 && tempo <= 183) {
```
Comment (sev-) [inline engines/director/score.cpp:502]: Digital videos were introduced in D3.1 or so. What does happen in earlier versions of Direcor?

File: engines/director/score.cpp [hunk 3]
```diff
@@ -462,15 +471,16 @@ void Score::update() {
...
+	// puppetTempo is overridden by changes in score tempo
+	if (_frames[_currentFrame]->_tempo || tempo != _lastTempo) {
 		_puppetTempo = 0;
 	} else if (_puppetTempo) {
 		tempo = _puppetTempo;
 	}
 
 	if (tempo) {
-		if (tempo > 161) {
+		if (tempo >= 196) {
```
Comment (sev-) [inline engines/director/score.cpp:483]: What does happen in D2 and D3?

File: video/video_decoder.cpp [hunk 4]
```diff
@@ -96,7 +96,26 @@ bool VideoDecoder::loadFile(const Common::Path &filename) {
 }
 
 bool VideoDecoder::needsUpdate() const {
-	return hasFramesLeft() && getTimeToNextFrame() == 0;
+	bool hasVideo = false;
+	bool hasAudio = false;
+	for (TrackList::const_iterator it = _tracks.begin(); it != _tracks.end(); it++) {
```
Comment (sev-) [inline video/video_decoder.cpp:101]: use `auto`?

**Review by sev- — COMMENTED** [review]
I am a bit concerned about tempo values. May it happen that in earlier versions of D they had different meanings? IIRC, the current values were take from D3 docs.

File: engines/director/score.cpp (hunk 3 above)
Comment (moralrecordings) [reply engines/director/score.cpp]: D2 and D3 allow delays of up to 120 seconds, D4 reduces this to 60,  have fixed with a conditional.

File: video/video_decoder.cpp (hunk 4 above)
Comment (moralrecordings) [reply video/video_decoder.cpp]: Fixed

File: engines/director/debugger.cpp (hunk 1 above)
Comment (moralrecordings) [reply engines/director/debugger.cpp]: Fixed

File: engines/director/score.cpp [hunk 5]
```diff
@@ -462,15 +471,18 @@ void Score::update() {
...
 		_puppetTempo = 0;
 	} else if (_puppetTempo) {
 		tempo = _puppetTempo;
 	}
 
 	if (tempo) {
-		if (tempo > 161) {
+		const bool waitForClickOnly = _vm->getVersion() < 300;
+		const int maxDelay = _vm->getVersion() < 400 ? 120 : 60;
+		if (tempo >= 256 - maxDelay) {
```
Comment (sev-) [inline engines/director/score.cpp:485]: Thank you for your investigation. However, this became complex and fragile. What do you think about adding a small table here enumerating the tempo ranges for every version?

**Comment by sev-** [comment]
Thank you! I am merging, but please add a table with the tempo constants, so we do not break it in the future.

If you created some test movies for the tempo, it would be great to have them added to the `director-tests` repo.

**Pull request merged**